### ISSUE TITLE
Tab completion everywhere

### ISSUE THREAD

**Issue by TylerLeonhardt** [PowerShell/SecretManagement#35]
This should work:

```pwsh
Get-SecretInfo -Vault <TAB>
Get-Secret -Name <TAB>
```

And complete based on the vaults that are available. Should be very simple to do with an argument completer.

**Comment by PaulHigin** [comment]
On my to do list :)

**Comment by TylerLeonhardt** [comment]
I also noticed that `Get-SecretInfo` has a `-Vault` but `Get-Secret` has `-VaultName`. That should probably be consistent.

**Comment by PaulHigin** [comment]
?? Both of these cmdlets have the parameter 'Vault'.  There is no cmdlet with 'VaultName'.
```powershell
Get-Command -ParameterName *vaultname*

```

**Comment by TylerLeonhardt** [comment]
I realize where my confusion was... It's here:

https://github.com/PowerShell/SecretManagement#testvaultextensionpsm1-file

The extension takes in `-VaultName` but the SM cmdlets take in `-Vault`. That's kinda confusing. Could they be consistent?

**Comment by PaulHigin** [comment]
Oh I see.  Yes this is inconsistent.  But I don't want to create yet another breaking change for vault authors for something like this.

**Comment by PaulHigin** [comment]
Closing issue as answered.

**Comment by SteveL-MSFT** [comment]
Removing `wontfix` label as that applies to `-Vault` vs `-VaultName`, but this issue is really about tab completion.  Tab completion should also apply to the name of the secret: `get-secret -name <tab>`

**Comment by TylerLeonhardt** [comment]
updated initial description

**Comment by itfranck** [comment]
I made my own little module a while ago (I need to review it for 0.5.3).
[SecretManagementArgumentCompleter](https://www.powershellgallery.com/packages/SecretManagementArgumentCompleter/2.0.0

I concur it would be super great to have that by default.
It is so convenient !

In my implementation, I create a local copy of the values obtained by `Get-SecretInfo` the first time and keep it on disk. 
Otherwise it takes 2-3 seconds with Azure key vaults (and potentially other remote providers) and this is far less practical. 

I added a `Clear-SecretManagementArgumentCompleterCache` to clear the file on disk.

Since this is not part of the official module (which would be awesom), I import the module in my profile so I don't have to think about it and have it ready to complete.

**Comment by itfranck** [comment]
Another thing that would be great, if there's an elegant way to do it, would be to have tab completion for the SecretVaultParameter. Implementing module would provide a method called `Get-SecretVaultModuleParameters` with the usual parameter set available so that you don't have to search implementing documentation every time. 

For instance, in my SecretManagementArgumentCompleter (this is a mouthful) module, I added the following: 
![image](https://user-images.githubusercontent.com/1980296/96352084-9ff58400-108e-11eb-9388-884f70524f18.png)

For me, this is great because I don't have to think about the name of theses every time. 
(It is hardcoded so if Az.Keyvault change the implementation, I need to update my module too) 
But if the information was provided by an implementing module, then it could be naturally filled out. 

Consider something like this:
```
Function Get-SecretVaultModuleParameters{
    [CmdletBinding()]
    param (
        [String]$ModuleName
    )
    $Context = [Microsoft.Azure.Commands.Common.Authentication.Abstractions.AzureRmProfileProvider]::Instance.Profile.DefaultContext    
    $SubscriptionId =  if ($null -ne $Context) { $Context.Subscription.Id } else { "" }
    return @( 
        # [SecretVaultParameterSet]::New()
        [PSCustomObject]@{
            Value    = @{ AZKVaultName = ''; SubscriptionId = $SubscriptionId }
            DisplayText = 'AzKeyVault / SubscriptionId '
            Description = 'AzKeyVault Name and subscription ID of the Keyvault'
        },
        # [SecretVaultParameterSet]::New()
        [PSCustomObject]@{
            Value       = @{ AZKVaultName = ''; SubscriptionId = $SubscriptionId;SomeOtherParam = 'Something' }
            DisplayText = 'AzKeyVault / SubscriptionId / SomeOtherParam '
            Description = 'AzKeyVault Name, subscription ID of the Keyvault and this other param'
        }
    )
}
```

The intellisense could use the parameter to build out a CompletionResult of type ParameterValue (so we get the tooltip)
- Value : ListItem (Parameter hashtable)
- DisplayText: CompletionText (This is what the user see in the completion result) 
- Description: Tooltip

---------------------

Using the AzKeyvault, I'd have completion for the required parameters (VaultName + SubscriptionId)
Using LastPass module, I'd see there is a **lpassCommand** parameter and it's description directly from the module.

tl;dr
Completion for the Vault parameter should be possible one way or another since it is a requirement for some vault and they won't work without it. Having NOT to rely on the documentation to apply parameters at registration time is great.

**Comment by PaulHigin** [comment]
Implemented for preview5.
